core/src/Cabana_ExecutionPolicy.hpp

```diff
@@ -13,53 +13,219 @@
 #define CABANA_EXECUTIONPOLICY_HPP
 
 #include <Cabana_Macros.hpp>
+#include <impl/Cabana_Index.hpp>
+
+#include <Kokkos_Core.hpp>
+
+#include <type_traits>
 
 namespace Cabana
 {
-namespace Experimental
-{
 //---------------------------------------------------------------------------//
 /*!
   \class RangePolicy
-  \brief Execution policy over a range of indices.
+  \brief Execution policy over a range of 1d indices.
+
+  Gives a linear range of indices for executing a functor with concurrency
+  over all indices.
 */
-template<int N, class ExecutionSpace>
-class RangePolicy
+template<class ExecutionSpace>
+class RangePolicy : public Kokkos::RangePolicy<ExecutionSpace>
 {
   public:
 
-    static constexpr int array_size = N;
+    using base_type = Kokkos::RangePolicy<ExecutionSpace>;
 
     using execution_space = ExecutionSpace;
 
-    // Range constructor.
+    /*!
+      \brief Range constructor.
+      \param begin The begininning of the 1D range.
+      \param begin The ending of the 1D range.
+    */
     RangePolicy( const std::size_t begin, const std::size_t end )
-        : _begin( begin )
-        , _end( end )
+        : base_type(begin,end)
+    {}
+
+    /*!
+      \brief Size constructor. Loop over n elements starting at 0.
+      \param n The number of elements in the range.
+    */
+    RangePolicy( const std::size_t n )
+        : base_type(0,n)
     {}
 
-    // Container constructor. The container must have a size() function that
-    // returns an int. C++ concepts would be really nice here. Valid
-    // containers include the AoSoA.
+    /*!
+      \brief Container constructor.
+
+      The container must have a size() function that returns an int. Valid
+      containers include the AoSoA, Slice, and Kokkos::View.
+
+      \param container The container over which to build the range policy.
+    */
     template<class Container>
     RangePolicy( Container container )
-        : _begin( 0 )
-        , _end( container.size() )
+        : base_type(0,container.size())
+    {}
+};
+//---------------------------------------------------------------------------//
+namespace Impl
+{
+/*!
+  \class
+
+  \brief 2D loop outer index range giving struct index range bounds based on a
+  1D range input.
+
+  \tparam VectorLength The inner array size of the AoSoA.
+*/
+template<int VectorLength>
+class StructRange
+{
+  public:
+
+    template<typename I>
+    KOKKOS_FORCEINLINE_FUNCTION
+    static constexpr
+    typename std::enable_if<std::is_integral<I>::value,std::size_t>::type
+    structBegin( const I& begin )
```

Comment:
Also really don't see the point of templating all these member functions.  It introduces unnecessary code bloat.  Just commit to an integral type, `std::size_t` is probably fine, otherwise if you really want a template, template the entire class on `IndexType` and instantiate it on `ExecutionSpace::index_type` or deduce it from the Kokkos policies using whatever traits they come with.

Reply:
Removed templates.